pkg/controllers/secret/secret_controller.go

```diff
@@ -0,0 +1,136 @@
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/multiClusterWatch"
+)
+
+// SecretReconciler reconciles a Secret object
+type SecretReconciler struct {
+	client.Client
+	Scheme  *runtime.Scheme
+	MCWatch multiClusterWatch.Interface
+}
+
+type TLSClientConfig struct {
+	Insecure bool   `json:"insecure"`
+	CaData   []byte `json:"caData,omitempty"`
+	CertData []byte `json:"certData,omitempty"`
+	KeyData  []byte `json:"keyData,omitempty"`
+}
+
+type ProviderConfig struct {
+	Command    string   `json:"command,omitempty"`
+	Args       []string `json:"args,omitempty"`
+	APIVersion string   `json:"apiVersion,omitempty"`
+}
+
+type ArgoClusterConfig struct {
+	BearerToken        string          `json:"bearerToken,omitempty"`
+	Username           string          `json:"username,omitempty"`
+	Password           string          `json:"password,omitempty"`
+	TlsClientConfig    TLSClientConfig `json:"tlsClientConfig,omitempty"`
+	ExecProviderConfig ProviderConfig  `json:"execProviderConfig,omitempty"`
+}
+
+const (
+	ARGO_CLUSTER_LABEL       = "argocd.argoproj.io/secret-type"
+	ARGO_CLUSTER_LABEL_VALUE = "cluster"
+)
+
+//+kubebuilder:rbac:groups=,resources=secret,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=,resources=secret/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=,resources=secret/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Secret object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.1/pkg/reconcile
+func (r *SecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+	previous := &corev1.Secret{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, previous)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	secret := previous.DeepCopy()
+
+	//only interested in ArgoCD cluster secrets
+	if !v1.HasLabel(secret.ObjectMeta, ARGO_CLUSTER_LABEL) || secret.ObjectMeta.Labels[ARGO_CLUSTER_LABEL] != ARGO_CLUSTER_LABEL_VALUE {
+		return ctrl.Result{}, nil
+	}
+
+	log.Log.Info("got Argo cluster secret", "secret", secret.Namespace+"/"+secret.Name, "server", string(secret.Data["server"]))
+
+	clusterConfig := &ArgoClusterConfig{}
+	err = json.Unmarshal(secret.Data["config"], clusterConfig)
```

Comment:
any reason this couldn't be unmarshalled into a rest.Config. All the fields seem to match and then just need to set the host?